ranking/src/ts_system/mod.rs

```diff
@@ -0,0 +1,314 @@
+        ret = f64::max(ret,
+                       f64::max(
+                           f64::max(f64::abs(RefCell::borrow(&a[i]).0.mu - b[i].0.mu),
+                                    f64::abs(RefCell::borrow(&a[i]).0.sigma - b[i].0.sigma)),
+                           f64::max(f64::abs(RefCell::borrow(&a[i]).1.mu - b[i].1.mu),
+                                    f64::abs(RefCell::borrow(&a[i]).1.sigma - b[i].1.sigma)),
+                       ));
+    }
+
+    ret
+}
+
+impl TrueSkillSPBSystem {
+    fn inference(&mut self, contest: &TSContest) {
+        if contest.is_empty() {
+            return;
+        }
+
+        // could be optimized, written that way for simplicity
+        let mut s = gen_player_message(contest, &ProdNode::new());
+        let mut perf = gen_player_message(contest, &ProdNode::new());
+        let mut p = gen_player_message(contest, &ProdNode::new());
+        let mut t = gen_team_message(contest, &ProdNode::new());
+        let mut u = gen_team_message(contest, &LeqNode::new(self.eps));
+        let mut l = vec![ProdNode::new(); contest.len()];
+        let mut d = vec![GreaterNode::new(2. * self.eps); contest.len() - 1];
+        let mut sp = Vec::new();
+        let mut pt = Vec::new();
+        let mut tul = Vec::new();
+        let mut ld = Vec::new();
+        let mut players = Vec::new();
+        let mut conv = Vec::new();
+        let mut old_conv = Vec::new();
+
+        for i in 0..contest.len() {
+            for j in 0..contest[i].len() {
+                for k in 0..contest[i][j].len() {
+                    players.push((contest[i][j][k].clone(), s[i][j][k].add_edge()));
+                    RefCell::borrow_mut(&players.last().unwrap().1.upgrade().unwrap()).0 =
+                        self.rating.get(&players.last().unwrap().0).unwrap().clone();
+
+                    let mut tmp: Vec<&mut dyn ValueNode> = Vec::with_capacity(3);
+                    tmp.push(&mut p[i][j][k]);
+                    tmp.push(&mut s[i][j][k]);
+                    tmp.push(&mut perf[i][j][k]);
+                    sp.push(SumNode::new(&mut tmp));
+                    RefCell::borrow_mut(perf[i][j][k].get_edges_mut().last_mut().unwrap()).1 = Gaussian { mu: 0., sigma: self.beta };
+                }
+
+                let mut tt: Vec<&mut dyn ValueNode> = vec![&mut t[i][j]];
+                for pp in &mut p[i][j] {
+                    tt.push(pp);
+                }
+                pt.push(SumNode::new(&mut tt));
+                let mut tmp: Vec<&mut dyn ValueNode> = Vec::with_capacity(3);
+                tmp.push(&mut l[i]);
+                tmp.push(&mut t[i][j]);
+                tmp.push(&mut u[i][j]);
+                tul.push(SumNode::new(&mut tmp));
+                conv.push(t[i][j].get_edges().last().unwrap().clone());
+            }
+
+            if i != 0 {
+                let mut tmp: Vec<&mut dyn ValueNode> = Vec::with_capacity(3);
+                let (a, b) = l.split_at_mut(i);

```

Comment:
This seems like a funny way to use `split_at_mut()`... why not directly use `l[i-1]` and `l[i]`?
```
ld.push(SumNode::new(&mut [
    &mut a[i-1],
    &mut a[i],
    &mut d[i-1]
]));
```

Reply:
`error[E0499]: cannot borrow `l` as mutable more than once at a time`

Reply:
Oops now I understand. Oof that's slightly annoying eh. Here's another hack then, use whatever you prefer:
```
let mut tmp: Vec<&mut dyn ValueNode> = l[i-1..=i].iter_mut().collect();
tmp.push(&mut d[i-1]);
```

Reply:
Or if you're concerned about the possible `Vec` re-allocation from `push()`, I guess the following could work too though maybe this is even more cryptic than what you had before:
```
let mut tmp: Vec<&mut dyn ValueNode> = l[i-1..=i].iter_mut().chain(Some(&mut d[i-1])).collect();
```